src/test/function_test/test_throttle.cpp

```diff
@@ -0,0 +1,785 @@
+        system("./run.sh clear_onebox");
+        system("./run.sh start_onebox -w");
+        chdir(global_env::instance()._working_dir.c_str());
+    }
+
+    void set_throttle(throttle_type type, uint64_t value)
+    {
+        std::vector<std::string> keys, values;
+        if (type == throttle_type::read_by_qps) {
+            keys.emplace_back("replica.read_throttling");
+            values.emplace_back(fmt::format("{}*reject*200", value));
+        } else if (type == throttle_type::read_by_size) {
+            keys.emplace_back("replica.read_throttling_by_size");
+            values.emplace_back(fmt::format("{}", value));
+        } else if (type == throttle_type::write_by_qps) {
+            keys.emplace_back("replica.write_throttling");
+            values.emplace_back(fmt::format("{}*reject*10", value));
+        } else if (type == throttle_type::write_by_size) {
+            keys.emplace_back("replica.write_throttling_by_size");
+            values.emplace_back(fmt::format("{}*reject*200", value));
+        }
+        auto resp = ddl_client->set_app_envs(app_name, keys, values);
+        dassert_f(
+            resp.get_error().code() == ERR_OK, "Set env failed: {}", resp.get_value().hint_message);
+    }
+
+    void restore_throttle()
+    {
+        std::map<std::string, std::string> envs;
+        ddl_client->get_app_envs(app_name, envs);
+        std::vector<std::string> keys;
+        for (const auto &iter : envs) {
+            keys.emplace_back(iter.first);
+        }
+        auto resp = ddl_client->del_app_envs(app_name, keys);
+        dassert_f(resp == ERR_OK, "Del env failed");
+    }
+
+    throttle_test_recorder start_test(throttle_test_plan test_plan, uint64_t time_duration_s = 10)
+    {
+        std::cout << fmt::format("start test, on {}", test_plan.test_plan_case) << std::endl;
+
+        dassert(pg_client, "pg_client is nullptr");
+
+        throttle_test_recorder r;
+        r.start_test(test_plan.test_plan_case, time_duration_s);
+
+        bool is_running = true;
+        std::atomic<int64_t> ref_count(0);
+
+        while (!r.is_time_up()) {
+            auto h_key = generate_hash_key_with_hotkey(test_plan.is_hotkey, 75, test_hashkey_len);
+            auto s_key = generate_random_str(test_sortkey_len);
+            auto value = generate_random_str(test_plan.random_value_size
+                                                 ? dsn::rand::next_u32(test_plan.single_value_sz)
+                                                 : test_plan.single_value_sz);
+            auto sortkey_value_pairs = generate_sortkey_value_map(
+                generate_str_vector_by_random(test_sortkey_len, test_plan.multi_count),
+                generate_str_vector_by_random(
+                    test_plan.single_value_sz, test_plan.multi_count, test_plan.random_value_size));
+
+            if (test_plan.ot == operation_type::set) {
+                ref_count++;
+                pg_client->async_set(
```

Comment:
I think you can encapsulate different operation into several small functions

Reply:
It's difficult to capture all the related parameters like `throttle_test_recorder` `is_running` `ref_count`